### PR TITLE
Add compatibility with monolog v3

### LINKED ISSUE #1318: Monolog handler is not compatible with monolog v3
[`Sentry\Monolog\Handler`](https://github.com/getsentry/sentry-php/blob/master/src/Monolog/Handler.php) extends [`Monolog\Handler\AbstractProcessingHandler`](https://github.com/Seldaek/monolog/blob/main/src/Monolog/Handler/AbstractProcessingHandler.php). 

The signature of the constructor has changed in [monolog 3](https://github.com/Seldaek/monolog/blob/3.0.0/UPGRADE.md#300) which leads to a fatal error:

```
Fatal error: Declaration of Sentry\Monolog\Handler::write(array $record): void must be compatible with Monolog\Handler\AbstractProcessingHandler::write(Monolog\LogRecord $record): void in /var/www/html/vendor/sentry/sentry/src/Monolog/Handler.php on line 51
```

### PR DESCRIPTION
**Pull request by W0rma**
Fixes #1318

### CODE REVIEW

**Comment by W0rma** [comment]
I'm a bit unsure on how to fix the phpstan and psalm pipeline.

psalm seems to analyse the code based on PHP 7.2 and I don't know how to change that.

The phpstan errors are related to the BC layer which makes sure that the handler can be used in monolog v2 and v3. Should I add the reported "errors" to the baseline file?

File: tests/Monolog/RecordFactory.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+namespace Sentry\Tests\Monolog;
+
+use DateTime;
+use DateTimeImmutable;
+use Monolog\Logger;
+use Monolog\LogRecord;
+
+class RecordFactory
+{
+    public static function create($message, $level, string $channel, array $context, array $extra, ?DateTime $datetime = null)
```
Comment (ste93cry) [inline tests/Monolog/RecordFactory.php:14]: Why `$message` has no typehint? Also, it looks like we never use the `$datetime` argument when calling this function, so we can remove it

File: tests/Monolog/RecordFactory.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+declare(strict_types=1);
+
+namespace Sentry\Tests\Monolog;
+
+use DateTime;
+use DateTimeImmutable;
+use Monolog\Logger;
+use Monolog\LogRecord;
+
+class RecordFactory
```
Comment (ste93cry) [inline tests/Monolog/RecordFactory.php:12]: Please make this class `@final` and mark it as `@internal`

**Review by ste93cry — CHANGES_REQUESTED** [review]
Thank you for the contribution. Since this is not a bug fix, please target the `develop` branch.

>psalm seems to analyse the code based on PHP 7.2 and I don't know how to change that.

Technically speaking, it's correct that Psalm complains because the code should run on any PHP version down to `7.2`. I think that the only viable solution is to update the baseline and ignore the errors. A more appropriate way could be to run the analysis more than once, one for PHP `7.2` and one for PHP `8.0` though, but it would be a bigger change and I wouldn't tackle it here

>The phpstan errors are related to the BC layer which makes sure that the handler can be used in monolog v2 and v3. Should I add the reported "errors" to the baseline file?

The error is because PHPStan does not understand that the `Logger::API` constant can change depending on which version of the package is installed. It should be enough to instruct it that the constant is dynamic by using the [`dynamicConstantNames`](https://phpstan.org/config-reference#constants) option.

File: src/Monolog/Handler.php [hunk 3]
```diff
@@ -48,24 +49,38 @@ public function __construct(HubInterface $hub, $level = Logger::DEBUG, bool $bub
     /**
      * {@inheritdoc}
      */
-    protected function write(array $record): void
+    protected function doWrite($record): void
     {
+        if (Logger::API >= 3) {
```
[the remark this answers is not in the document]
Comment (W0rma) [reply src/Monolog/Handler.php]: Thank you for the hint.

**Comment by W0rma** [comment]
@ste93cry Thank you for the hints! I have changed the target branch to `develop` and addressed all your suggestions.

File: src/Monolog/CompatibilityProcessingHandlerTrait.php [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            $this->doWrite($record);
+        }
+
+        /**
+         * Translates the Monolog level into the Sentry severity.
+         */
+        private static function getSeverityFromLevel(int $level): Severity
+        {
+            $level = Level::from($level);
```
Comment (W0rma) [inline src/Monolog/CompatibilityProcessingHandlerTrait.php:38]: I tried to use `Level::Debug->value` in the `switch` statement but this broke the tests in PHP < 8.0.

**Review by Jean85 — CHANGES_REQUESTED** [review]
Good work! Please add a changelog entry.

**Comment by W0rma** [comment]
> Good work! Please add a changelog entry.

@jean85 Done

**Review by Jean85 — APPROVED** [review]

File: src/Monolog/CompatibilityProcessingHandlerTrait.php (hunk 4 above)
Comment (ste93cry) [reply src/Monolog/CompatibilityProcessingHandlerTrait.php:38]: Makes definitely sense: we could not use anyway the PHP syntax for a newer version in a file that is also used in a previous version

File: tests/Monolog/RecordFactory.php (hunk 1 above)
Comment (W0rma) [reply tests/Monolog/RecordFactory.php]: > Why $message has no typehint?

I have just added the missing type hints.

**Review by ste93cry — CHANGES_REQUESTED** [review]
I've left a few more comments, but they are just nitpick. You should also rebase the PR to solve the issues with PHPStan errors. The only concern I have, but to which I don't think there is solution that can be easily taken here, is that it looks weird to see errors related to the `vendor` folder in the Psalm baseline 😞

File: src/Monolog/CompatibilityProcessingHandlerTrait.php [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+         */
+        protected function write(array $record): void
+        {
+            $this->doWrite($record);
+        }
+
+        /**
+         * Translates the Monolog level into the Sentry severity.
+         *
+         * @param Logger::DEBUG|Logger::INFO|Logger::NOTICE|Logger::WARNING|Logger::ERROR|Logger::CRITICAL|Logger::ALERT|Logger::EMERGENCY $level The Monolog log level
```
[the remark this answers is not in the document]
Comment (W0rma) [reply src/Monolog/CompatibilityProcessingHandlerTrait.php:82]: I guess you refer to the type defined in https://github.com/Seldaek/monolog/blob/2.6.0/src/Monolog/Logger.php#L30. This definition is available in monolog v2 but was removed in monolog v3.

Should I use it anyway?

File: composer.json [hunk 6]
```diff
@@ -41,7 +41,7 @@
     "require-dev": {
         "friendsofphp/php-cs-fixer": "^2.19|3.4.*",
         "http-interop/http-factory-guzzle": "^1.0",
-        "monolog/monolog": "^1.3|^2.0",
+        "monolog/monolog": "^1.6|^2.0|^3.0",
```
[the remark this answers is not in the document]
Comment (W0rma) [reply composer.json:44]: I wanted to make sure that the constant `Monolog\Logger::API` is available which was added in 1.6 (see https://github.com/Seldaek/monolog/blob/main/CHANGELOG.md#160-2013-07-29).

[the remark this answers is not in the document]
Comment (ste93cry) [reply composer.json:44]: Ok, definitely makes sense!

File: src/Monolog/CompatibilityProcessingHandlerTrait.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (ste93cry) [reply src/Monolog/CompatibilityProcessingHandlerTrait.php:82]: No it’s ok, I thought it was available in all versions

**Review by ste93cry — APPROVED** [review]
Thank you for the hard work!

**Pull request merged**